File: 剑指offer/字符串的排列.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    vector <string> res;
    vector<string> Permutation(string str) {
        if (str.empty() || 0 == str.length())
            return res;
        PermuHelper(str, 0);
        // 排序
        sort(res.begin(), res.end());
        return res;
    }
    void PermuHelper(string str, int start) {
        if (start == str.length() - 1)
            if (res.end() == std :: find(res.begin(), res.end(), str))
                res.push_back(str);
        for (int i = start; i < str.length(); ++i) {
            swapEl(str, start, i);
            PermuHelper(str, start + 1);
            swapEl(str, start, i);
        }
    }
    void swapEl(string & str, int i, int j) {
        char temp = str[i];
        str[i] = str[j];
        str[j] = temp;
    }
};

int main ()
{
    vector <string> res;
    auto solve = new Solution();
    res = solve -> Permutation("abc");
    for (int i = 0; i < res.size(); ++i)
        cout << res[i] << endl;
    return 0;
}

File: 剑指offer/字符串的排列.js
function Permutation(str)
{
  // write code here
  let res = []
  if (!str.length)  return res
  arr = str.split('') //  转化为字符串数组
  res = Permutate(arr, 0, res)
  res = [...new Set(res)] //  去重
  res.sort()  //  排序
  return res
}

function Permutate (arr, index, res) {
  if (arr.length === index) {
    let str = ''
    for (let i = 0; i < arr.length; i++)
      str += arr[i]
    return res.push(str)
  }
  for (let i = index; i < arr.length; i++) {
    [arr[index], arr[i]] = [arr[i], arr[index]] // 交换
    Permutate(arr, index + 1, res)
    [arr[index], arr[i]] = [arr[i], arr[index]] // 交换
  }
  return res
}

const res = Permutation('ab')
console.log(res)
